File: 2022_python/220403.py
#
# created by z on 4/3/2022
# 31

class Solution:
    def nextPermutation(self, nums: List[int]) -> None:
        """
        Do not return anything, modify nums in-place instead.
        """
        n: int = len(nums)
        if n == 1:
            return

        op: int = n - 1
        while op > 0 and nums[op - 1] >= nums[op]:
            op -= 1

        if op != 0:
            ex: int = op
            while ex < n and nums[op - 1] < nums[ex]:
                ex += 1

            nums[op - 1], nums[ex - 1] = nums[ex - 1], nums[op - 1]

        nums[op:] = sorted(nums[op:])
        return

File: 2022/2022_04_cpp/220403.cpp
//
// Created by z on 4/3/2022.
// # 31

class Solution {
public:
    void nextPermutation(vector<int>& nums) {
        if (nums.size() == 1)
            return;

        int op_idx = nums.size() - 1;
        while (op_idx > 0 && nums[op_idx-1] >= nums[op_idx])
            op_idx--;

        if(op_idx != 0)
        {
            int ex_idx = op_idx;
            while (ex_idx < nums.size() && nums[ex_idx] > nums[op_idx-1])
                ex_idx++;
            int tmp = nums[ex_idx-1];
            nums[ex_idx-1] = nums[op_idx-1];
            nums[op_idx-1] = tmp;
        }

        sort(nums.begin()+op_idx, nums.end());
        return;
    }
};

